### PR TITLE
[API] Shipping methods resolution

### PR DESCRIPTION
**Pull request by TheMilek**
| Q               | A
| --------------- | -----
| Branch?         | master 
| Bug fix?        | yes
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no 
| License         | MIT

ShippingMethods as an subresource was causing unwanted behaviour - we decided to make it into a collection to solve the problem

### CODE REVIEW

File: features/shipping/applying_shipping_method_rules/viewing_available_shipping_methods_based_on_channel_as_a_guest.feature [hunk 1]
```diff
@@ -0,0 +1,33 @@
+@applying_shipping_method_rules
+Feature: Viewing available shipping methods based on channel
```
Comment (AdamKasp) [inline features/shipping/applying_shipping_method_rules/viewing_available_shipping_methods_based_on_channel_as_a_guest.feature:2]: We are trying to name features in the same way as file name

File: features/shipping/applying_shipping_method_rules/viewing_available_shipping_methods_based_on_channel_as_an_logged_in_customer.feature [hunk 2]
```diff
@@ -0,0 +1,34 @@
+@applying_shipping_method_rules
+Feature: Viewing available shipping methods based on channel
```
Comment (AdamKasp) [inline features/shipping/applying_shipping_method_rules/viewing_available_shipping_methods_based_on_channel_as_an_logged_in_customer.feature:2]: same

Comment (AdamKasp) [reply features/shipping/applying_shipping_method_rules/viewing_available_shipping_methods_based_on_channel_as_an_logged_in_customer.feature:2]: can we merge these two features? both are related with available shipping methods

File: src/Sylius/Behat/Resources/config/services/api.xml [hunk 3]
```diff
@@ -24,6 +24,11 @@
             <argument>shop</argument>
         </service>
 
+        <service id="sylius.behat.api_platform_client.shop.shipping-methods" class="Sylius\Behat\Client\ApiPlatformClient" parent="sylius.behat.api_platform_client">
```
Comment (AdamKasp) [inline src/Sylius/Behat/Resources/config/services/api.xml:27]:
```suggestion
        <service id="sylius.behat.api_platform_client.shop.shipping_methods" class="Sylius\Behat\Client\ApiPlatformClient" parent="sylius.behat.api_platform_client">
```

File: src/Sylius/Bundle/ApiBundle/DataProvider/CartShippingMethodsCollectionDataProvider.php [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        ShippingMethodRepositoryInterface $shippingMethodsRepository,
+        ShippingMethodsResolverInterface $shippingMethodsResolver,
+        ChannelContextInterface $channelContext
+    ) {
+        $this->orderRepository = $orderRepository;
+        $this->shipmentRepository = $shipmentRepository;
+        $this->shippingMethodsRepository = $shippingMethodsRepository;
+        $this->shippingMethodsResolver = $shippingMethodsResolver;
+        $this->channelContext= $channelContext;
+    }
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/DataProvider/CartShippingMethodsCollectionDataProvider.php:54]:
```suggestion
    public function __construct(
        private OrderRepositoryInterface $orderRepository,
        private ShipmentRepositoryInterface $shipmentRepository,
        private ShippingMethodRepositoryInterface $shippingMethodsRepository,
        private ShippingMethodsResolverInterface $shippingMethodsResolver,
        private ChannelContextInterface $channelContext
    ) {
    }
```

File: src/Sylius/Bundle/ApiBundle/DataProvider/CartShippingMethodsCollectionDataProvider.php [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        if ($shipment === null) {
+            return [];
+        }
+
+        return $this->shippingMethodsResolver->getSupportedMethods($shipment);
+    }
+
+    public function supports(string $resourceClass, string $operationName = null, array $context = []): bool
+    {
+        return is_a($resourceClass, ShippingMethodInterface::class, true); // only for shop context
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/DataProvider/CartShippingMethodsCollectionDataProvider.php:90]: do we need this comment?

File: src/Sylius/Bundle/ApiBundle/Serializer/ShippingMethodNormalizer.php [hunk 6]
```diff
@@ -31,11 +34,24 @@ final class ShippingMethodNormalizer implements ContextAwareNormalizerInterface,
...
+    private ShipmentRepositoryInterface $shipmentRepository;
+
+    private ServiceRegistryInterface $shippingCalculators;
+
+    private RequestStack $requestStack;
+
     public function __construct(
-        private OrderRepositoryInterface $orderRepository,
-        private ShipmentRepositoryInterface $shipmentRepository,
-        private ServiceRegistryInterface $shippingCalculators
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/Serializer/ShippingMethodNormalizer.php:37]: why did you change it?

File: src/Sylius/Bundle/ApiBundle/Serializer/ShippingMethodNormalizer.php [hunk 7]
```diff
@@ -45,24 +61,33 @@ public function normalize($object, $format = null, array $context = [])
...
+        if (!isset($filters)) {
+            return $data;
+        }
+
+        Assert::keyExists($filters, 'tokenValue');
+        Assert::keyExists($filters, 'shipmentId');
 
         /** @var ShipmentInterface $shipment */
-        $shipment = $this->shipmentRepository->find($shipmentId);
+        $shipment = $this->shipmentRepository->find($filters['shipmentId']); // Duplication of logic from CartShippingMethodCollectionDataProvider
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/Serializer/ShippingMethodNormalizer.php:81]: do we need this comment?

File: features/shipping/applying_shipping_method_rules/viewing_available_shipping_methods_based_on_channel.feature [hunk 8]
```diff
@@ -0,0 +1,51 @@
+@applying_shipping_method_rules
```
Comment (lchrusciel) [inline features/shipping/applying_shipping_method_rules/viewing_available_shipping_methods_based_on_channel.feature:1]: Probably it would be better to introduce new namespace for these features
```suggestion
@viewing_shipping_methods
```

File: features/shipping/applying_shipping_method_rules/viewing_available_shipping_methods_based_on_channel.feature [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+Feature: Viewing available shipping methods based on channel
+    In order to only see applicable shipping methods
+    As a guest or as a logged in customer
```
Comment (lchrusciel) [inline features/shipping/applying_shipping_method_rules/viewing_available_shipping_methods_based_on_channel.feature:4]:
```suggestion
    As a Visitor or Shop user
```

File: features/shipping/applying_shipping_method_rules/viewing_available_shipping_methods_based_on_channel.feature [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    I want to see the shipping methods that are available to my order based on the channel
+
+    Background:
+        Given the store operates on a channel named "United Kingdom" in "USD" currency and with hostname "uk.cool-clothes.example"
+        And the store operates on another channel named "United States" in "USD" currency and with hostname "usa.cool-clothes.example"
+        And the store operates on another channel named "Colombia" in "USD" currency and with hostname "colombia.cool-clothes.example"
```
Comment (lchrusciel) [inline features/shipping/applying_shipping_method_rules/viewing_available_shipping_methods_based_on_channel.feature:10]: Not needed
```suggestion
```

File: features/shipping/applying_shipping_method_rules/viewing_available_shipping_methods_based_on_channel.feature [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+        And the store has "uber speedy" shipping method with "$15.00" fee per shipment for "United Kingdom" channel and "$12.00" for "Colombia" channel
+        And the store has a product "T-Shirt" priced at "$20.00" available in channel "United Kingdom" and channel "United States"
+
+    @api @ui
+    Scenario: Seeing shipping methods that are available in channel as a guest
+        Given I changed my current channel to "United States"
+        And I have product "T-Shirt" in the cart
+        When I complete addressing step with email "john@example.com" and "United States" based billing address
+        Then I should be on the checkout shipping step
+        And I should see "ultra fast" shipping method
```
Comment (lchrusciel) [inline features/shipping/applying_shipping_method_rules/viewing_available_shipping_methods_based_on_channel.feature:25]: Can we check that the second method is not visible

Comment (lchrusciel) [reply features/shipping/applying_shipping_method_rules/viewing_available_shipping_methods_based_on_channel.feature:25]: Same of the rest of steps

File: features/shipping/applying_shipping_method_rules/viewing_available_shipping_methods_based_on_channel.feature (hunk 8 above)
Comment (TheMilek) [reply features/shipping/applying_shipping_method_rules/viewing_available_shipping_methods_based_on_channel.feature:1]: you mean changing catalog's name also? There are other features starting with `viewing_shipping_methods`

Comment (TheMilek) [reply features/shipping/applying_shipping_method_rules/viewing_available_shipping_methods_based_on_channel.feature:1]: Zrzut ekranu 2022-02-10 o 16 54 35

File: src/Sylius/Bundle/ApiBundle/DataProvider/CartShippingMethodsCollectionDataProvider.php [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+    ) {
+    }
+
+    public function getCollection(string $resourceClass, string $operationName = null, array $context = []): array
+    {
+        /** @var ChannelInterface $channel */
+        $channel = $this->channelContext->getChannel();
+
+        if (!isset($context['filters'])) {
+
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/DataProvider/CartShippingMethodsCollectionDataProvider.php:47]:
```suggestion
```

File: features/shipping/viewing_shipping_methods/viewing_available_shipping_methods_based_on_channel.feature [hunk 13]
```diff
@@ -0,0 +1,54 @@
+@viewing_shipping_methods
+Feature: Viewing available shipping methods based on channel
+    In order to only see applicable shipping methods
+    As a Visitor or Shop user
+    I want to see the shipping methods that are available to my order based on the channel
+
+    Background:
+        Given the store operates on a channel named "United Kingdom" in "USD" currency and with hostname "uk.cool-clothes.example"
+        And the store operates on another channel named "United States" in "USD" currency and with hostname "usa.cool-clothes.example"
+        And the store has a zone "World" with code "WRLD"
```
Comment (GSadee) [inline features/shipping/viewing_shipping_methods/viewing_available_shipping_methods_based_on_channel.feature:10]:
```suggestion
        And the store has a zone "World"
```

File: features/shipping/viewing_shipping_methods/viewing_available_shipping_methods_based_on_channel.feature [hunk 14]
```diff
@@ -0,0 +1,54 @@
+@viewing_shipping_methods
+Feature: Viewing available shipping methods based on channel
+    In order to only see applicable shipping methods
+    As a Visitor or Shop user
```
Comment (GSadee) [inline features/shipping/viewing_shipping_methods/viewing_available_shipping_methods_based_on_channel.feature:4]:
```suggestion
    As a Visitor or Shop User
```

File: features/shipping/viewing_shipping_methods/viewing_available_shipping_methods_based_on_channel.feature [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    I want to see the shipping methods that are available to my order based on the channel
+
+    Background:
+        Given the store operates on a channel named "United Kingdom" in "USD" currency and with hostname "uk.cool-clothes.example"
+        And the store operates on another channel named "United States" in "USD" currency and with hostname "usa.cool-clothes.example"
+        And the store has a zone "World" with code "WRLD"
+        And the store ships to "United States"
+        And this zone has the "United States" country member
+        And the store ships everywhere for free for all channels
+        And the store has "ultra fast" shipping method with "$4" fee per unit for "United States" channel
```
Comment (GSadee) [inline features/shipping/viewing_shipping_methods/viewing_available_shipping_methods_based_on_channel.feature:14]:
```suggestion
        And the store has "ultra fast" shipping method with "$4.00" fee per unit for "United States" channel
```

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 16]
```diff
@@ -49,6 +49,8 @@ final class CheckoutContext implements Context
 
     private ApiClientInterface $addressesClient;
 
+    private ApiClientInterface $shippingMethodClient;
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:52]: To have consistent name with other clients:
```suggestion
    private ApiClientInterface $shippingMethodsClient;
```

File: src/Sylius/Behat/Context/Setup/ShippingContext.php [hunk 17]
```diff
@@ -258,6 +258,29 @@ public function storeHasShippingMethodWithFeePerShipmentForChannels(
         ]));
     }
 
+    /**
+     * @Given /^the store has "([^"]+)" shipping method with ("[^"]+") fee per shipment for ("[^"]+" channel)$/
+     */
+    public function storeHasShippingMethodWithFeePerShipmentForChannel(
+        $shippingMethodName,
+        $fee,
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Setup/ShippingContext.php:266]: Missing arguments types

File: src/Sylius/Bundle/ApiBundle/DataProvider/CartShippingMethodsCollectionDataProvider.php [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+use Sylius\Component\Shipping\Resolver\ShippingMethodsResolverInterface;
+use Webmozart\Assert\Assert;
+
+/** @experimental */
+final class CartShippingMethodsCollectionDataProvider implements ContextAwareCollectionDataProviderInterface, RestrictedDataProviderInterface
+{
+    public function __construct(
+        private OrderRepositoryInterface $orderRepository,
+        private ShipmentRepositoryInterface $shipmentRepository,
+        private ShippingMethodRepositoryInterface $shippingMethodsRepository,
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/DataProvider/CartShippingMethodsCollectionDataProvider.php:35]: Be consistent with naming:
```suggestion
        private ShippingMethodRepositoryInterface $shippingMethodRepository,
```

File: src/Sylius/Bundle/ApiBundle/DataProvider/CartShippingMethodsCollectionDataProvider.php (hunk 12 above)
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/DataProvider/CartShippingMethodsCollectionDataProvider.php:47]:
```suggestion
```

File: src/Sylius/Bundle/ApiBundle/Filter/ShippingMethodFilter.php [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+                    'description' => 'Id of shipment for which you would like to check available shipping methods',
+                ],
+            ],
+            'tokenValue' => [
+                'type' => 'string',
+                'required' => false,
+                'property' => null,
+                'swagger' => [
+                    'name' => 'Order',
+                    'description' => 'Id of order for which you would like to check available shipping methods. Should be set if you are checking shipping methods for shipment, that is not assigned to user.',
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Filter/ShippingMethodFilter.php:38]: That's not true, you are using token value of an order, not its id:
```suggestion
                    'description' => 'Token of an order for which you would like to check available shipping methods. Should be set if you are checking shipping methods for shipment, that is not assigned to user.',
```

File: src/Sylius/Bundle/ApiBundle/spec/Serializer/ShippingMethodNormalizerSpec.php [hunk 20]
```diff
@@ -14,43 +14,49 @@
...
     }
 
-    public function it_supports_only_shipping_method_interface(ShippingMethodInterface $shippingMethod, OrderInterface $order): void
+    public function it_supports_only_shipping_methods_interface_and_shop_context(ShippingMethodInterface $shippingMethod): void
     {
         $this
-            ->supportsNormalization($shippingMethod, null, ['subresource_identifiers' => ['tokenValue' => '666', 'shipments' => '999']])
+            ->supportsNormalization($shippingMethod, null, [
+                    'collection_operation_name' => 'shop_get',
+                    'filters' => ['tokenValue' => '666', 'shipmentId' => '999']
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/spec/Serializer/ShippingMethodNormalizerSpec.php:48]: Too big indent :dancer:

File: src/Sylius/Bundle/ApiBundle/spec/Serializer/ShippingMethodNormalizerSpec.php [hunk 21]
```diff
@@ -80,18 +86,29 @@ public function it_serializes_shipping_method_if_item_operation_name_is_differen
...
+        $channelContext->getChannel()->willReturn($channel);
+
+        $request->attributes = new ParameterBag(['_api_filters' => ['tokenValue' => '666', 'shipmentId' => '999']]);
+
+        $orderRepository->findCartByTokenValueAndChannel('666', $channel)->willReturn($cart);
+        $cart->getId()->willReturn('111');
+        $shipmentRepository->findOneByOrderId('999', '111')->willReturn($shipment);
         $cart->hasShipment($shipment)->willReturn(true);
 
+
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/spec/Serializer/ShippingMethodNormalizerSpec.php:105]: Extra blank line

**Review by AdamKasp — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/DataProvider/CartShippingMethodsCollectionDataProvider.php [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+        if ($shipment === null) {
+            return [];
+        }
+
+        return $this->shippingMethodsResolver->getSupportedMethods($shipment);
+    }
+
+    public function supports(string $resourceClass, string $operationName = null, array $context = []): bool
+    {
+        return is_a($resourceClass, ShippingMethodInterface::class, true) && $this->isShopGetCollectionOperation($context);
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/DataProvider/CartShippingMethodsCollectionDataProvider.php:63]:
```suggestion
        return $resourceClass === ShippingMethodInterface::class && $this->isShopGetCollectionOperation($context);
```

File: features/shipping/viewing_shipping_methods/viewing_available_shipping_methods_based_on_channel.feature [hunk 23]
```diff
@@ -0,0 +1,54 @@
+@viewing_shipping_methods
+Feature: Viewing available shipping methods based on channel
+    In order to only see applicable shipping methods
+    As a Visitor or Shop User
```
Comment (Zales0123) [inline features/shipping/viewing_shipping_methods/viewing_available_shipping_methods_based_on_channel.feature:4]: Do these scenarios differ for Visitors and Shop Users? If they do, cannot they be in two separate files? If they do not, do we need to check them for both of these actors?

File: src/Sylius/Bundle/ApiBundle/Serializer/ShippingMethodNormalizer.php [hunk 24]
```diff
@@ -45,24 +51,36 @@ public function normalize($object, $format = null, array $context = [])
 
         $context[self::ALREADY_CALLED] = true;
 
-        $subresourceIdentifiers = $context['subresource_identifiers'];
+        $request = $this->requestStack->getCurrentRequest();
 
-        $shipmentId = $subresourceIdentifiers['shipments'] ?? $subresourceIdentifiers['id'];
+        if (null === $filters = $request->attributes->get('_api_filters')) {
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/Serializer/ShippingMethodNormalizer.php:56]: Looks messy :dancer: How about assigning the variable before the condition?

```php
$filters = $request->attributes->get('_api_filters');
if (null === $filters) {
```

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ShipmentRepository.php [hunk 25]
```diff
@@ -41,6 +42,21 @@ public function findOneByOrderId($shipmentId, $orderId): ?ShipmentInterface
...
+            ->innerJoin('o.order', 'ord')
+            ->andWhere('o.id = :shipmentId')
+            ->andWhere('ord.tokenValue = :tokenValue')
+            ->andWhere('ord.channel = :channel')
+            ->setParameter('shipmentId', $shipmentId)
+            ->setParameter('tokenValue', $tokenValue)
+            ->setParameter('channel', $channel)
+            ->getQuery()
+            ->getOneOrNullResult()
+            ;
```
Comment (Zales0123) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ShipmentRepository.php:57]: Wrong indent :dancer:

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ShipmentRepository.php [hunk 26]
```diff
@@ -41,6 +42,21 @@ public function findOneByOrderId($shipmentId, $orderId): ?ShipmentInterface
         ;
     }
 
+    public function findOneByOrderTokenAndChannel($shipmentId, string $tokenValue, ChannelInterface $channel): ?ShipmentInterface
+    {
+        return $this->createQueryBuilder('o')
+            ->innerJoin('o.order', 'ord')
```
Comment (Zales0123) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ShipmentRepository.php:48]: `'ord'` -> `'order'`? 😅

File: features/shipping/viewing_shipping_methods/viewing_available_shipping_methods_based_on_channel.feature (hunk 23 above)
Comment (TheMilek) [reply features/shipping/viewing_shipping_methods/viewing_available_shipping_methods_based_on_channel.feature:4]: These checks 2 situations - which shipping methods see Shop User and which Visitor during the cart checkout. I think We can leave them in this one file, becouse the only difference for them is one step:
`And I am a logged in customer`

File: src/Sylius/Bundle/ApiBundle/Serializer/ShippingMethodNormalizer.php [hunk 27]
```diff
@@ -45,24 +51,37 @@ public function normalize($object, $format = null, array $context = [])
...
-        if (isset($subresourceIdentifiers['tokenValue'])) {
-            /** @var OrderInterface|null $cart */
-            $cart = $this->orderRepository->findCartByTokenValue($subresourceIdentifiers['tokenValue']);
-            Assert::notNull($cart);
+        Assert::keyExists($filters, 'tokenValue');
+        Assert::keyExists($filters, 'shipmentId');
 
-            Assert::true($cart->hasShipment($shipment), 'Shipment doesn\'t match for order');
-        }
+        /** @var ChannelInterface $channel */
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/Serializer/ShippingMethodNormalizer.php:71]: do we need this Normalizer? or maybe just provider will be enough

**Review by AdamKasp — CHANGES_REQUESTED** [review]

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ShipmentRepository.php (hunk 26 above)
Comment (TheMilek) [reply src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ShipmentRepository.php]: naming it `order` is causing problems when I try to implement it, probably it's restricted by `querybuilder`, what do you think about naming it for example `orders` instead?

File: src/Sylius/Bundle/ApiBundle/spec/DataProvider/CartShippingMethodsCollectionDataProviderSpec.php [hunk 28]
```diff
@@ -0,0 +1,180 @@
...
+                Request::METHOD_GET,
+                [
+                    'filters' => ['tokenValue' => '666', 'shipmentId' => ''],
+                ],
+            )
+            ->shouldReturn([])
+        ;
+    }
+
+    function it_returns_cart_shipping_methods(
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/spec/DataProvider/CartShippingMethodsCollectionDataProviderSpec.php:131]:
```suggestion
    function it_returns_shipping_methods_resolved_based_on_given_shipment_and_order(
```

File: src/Sylius/Bundle/ApiBundle/Serializer/ShippingMethodNormalizer.php [hunk 29]
```diff
@@ -45,24 +51,37 @@ public function normalize($object, $format = null, array $context = [])
...
+        if (null === $filters) {
+            return $data;
+        }
 
-        if (isset($subresourceIdentifiers['tokenValue'])) {
-            /** @var OrderInterface|null $cart */
-            $cart = $this->orderRepository->findCartByTokenValue($subresourceIdentifiers['tokenValue']);
-            Assert::notNull($cart);
+        Assert::keyExists($filters, 'tokenValue');
+        Assert::keyExists($filters, 'shipmentId');
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Serializer/ShippingMethodNormalizer.php:69]: I know, I suggested these assertions, but after second thought, can we just use early return here? The reasoning behind this is, that with these asserts we are cementing this endpoint and no one will ba able to easily add additional errors

File: tests/Api/DataFixtures/ORM/authentication/customer.yaml [hunk 30]
```diff
@@ -6,6 +6,13 @@ Sylius\Component\Core\Model\ShopUser:
         customer: "@customer_<current()>"
         username: "<current()>\\@doe.com"
         usernameCanonical: "<current()>\\@doe.com"
+    shop_user_{dave}:
```
Comment (lchrusciel) [inline tests/Api/DataFixtures/ORM/authentication/customer.yaml:9]: As this configuration is identical to the above one, we could simplify everything from lines 2 to 15 to: 
```yml
Sylius\Component\Core\Model\ShopUser:
    shop_user_{oliver, dave}:
        plainPassword: "sylius"
        roles: [ROLE_USER]
        enabled: true
        customer: "@customer_<current()>"
        username: "<current()>\\@doe.com"
        usernameCanonical: "<current()>\\@doe.com"
```

Ref. https://github.com/nelmio/alice/blob/master/doc/complete-reference.md#fixture-lists

File: tests/Api/DataFixtures/ORM/authentication/customer.yaml [hunk 31]
```diff
@@ -14,3 +21,9 @@ Sylius\Component\Core\Model\Customer:
         email: "<current()>\\@doe.com"
         emailCanonical: "<current()>\\@doe.com"
         birthday: "<(new \\DateTime())>"
+    customer_{dave}:
```
Comment (lchrusciel) [inline tests/Api/DataFixtures/ORM/authentication/customer.yaml:24]: same as above

**Review by AdamKasp — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/Serializer/ShippingMethodNormalizer.php (hunk 29 above)
Comment (TheMilek) [reply src/Sylius/Bundle/ApiBundle/Serializer/ShippingMethodNormalizer.php:69]: Do you think I should  also change `notNull` asserts on `$cart` and `$shipment`?

Comment (lchrusciel) [reply src/Sylius/Bundle/ApiBundle/Serializer/ShippingMethodNormalizer.php]: It is fine, as it is

**Comment by lchrusciel** [comment]
Thank you, Kamil! :tada:
